week1-webclient_unittests.md

```diff
@@ -0,0 +1,142 @@
+    - The line `const { getByText } = render(<PetCard pet={petMock}/>);` renders the PetCard using the mock data defined in 6 above, and returns a function `getByText` that allows you to check whether certain text values exist in the rendered component.
+    - The line `expect(getByText('Fido')).toBeInTheDocument();` defines an expectation that the text `Fido` should be in the rendered component.
+
+8. There are many different ways in which the rendered component and it's parts can be checked for validity.
+
+    - There are many ways in which sub parts of the component can be discovered in addition to `getByText` used above. See the [dom-testing-library cheatsheet](https://testing-library.com/docs/dom-testing-library/cheatsheet#queries) for a list.
+    - There are also many ways in which checks can be done on these discovered sub parts in addition to `toBeInTheDocument` used above. There are checks to do anything from whether an attribute exists on a dom element to whether an element is visible or not. See the [jest-dom documentation](https://www.npmjs.com/package/jest-dom#custom-matchers) for more examples.
+
+9. See if you can add additional tests for the following expectations:
+
+        PetCard should render the given description
+        PetCard should render the adopt button
+
+10. See the bottom of this page for an end to end solution containing one possible way to write these.
+
+### Running the tests.
+
+To run the tests execute the following in a terminal:
+
+    npm run test
+
+This will start a process that runs continuously. The process will recompile the code if anything changes and is saved, and also re-run the tests at that time. The results that are printed out looks similar to this when all the tests succeeded:
+
+    Test Suites: 2 passed, 2 total
+    Tests:       4 passed, 4 total
+
+Each seperate test file is considered a test suite and each `it` is counted a single test.
+
+Test failures are printed out on the console as well, e.g:
+
+    Test Suites: 1 failed, 1 passed, 2 total
+    Tests:       1 failed, 3 passed, 4 total
+
+To simulate a failure, try changing the unit test described above, so that the text that is expected is `Spot` instead of `Fido`, e.g:
+
+    expect(getByText('Spot')).toBeInTheDocument();
+
+This will print out detailed information about the test that failed including the expectation that wasn't met.
+
+### Best practices when writing tests.
+
+The description that you give your tests should be similar to a specification. A good template to consider is `Code a should do action b in case of c`. The code in the test should of do checks that match the description. Since requirements tend to change over time or even get lost, unit tests can serve as an accurate and up to date specification of the expected behavior of the application. Since they are checked in with the code, they are easier to keep in sync with the code and maintain than purely textual specifications.
+
+### Full sample solution.
```

Comment:
Suggestion: enclose this into `<details><summary></summary><details>` tags as I did for https://github.com/codessintheclassroom/classroom-material/blob/master/week1-webclient_ui.md mostly for coherence of style

Reply:
Oh yeah, that's nifty, updated to use it.